action/protocol/poll/slasher.go

```diff
@@ -544,11 +544,23 @@ func (sh *Slasher) updateCurrentBlockMeta(ctx context.Context, sm protocol.State
 }
 
 // calculateBlockProducer calculates block producer by given candidate list
-func (sh *Slasher) calculateBlockProducer(candidates state.CandidateList) (state.CandidateList, error) {
-	var blockProducers state.CandidateList
+func (sh *Slasher) calculateBlockProducer(ctx context.Context, candidates state.CandidateList) (state.CandidateList, error) {
+	var (
+		blockProducers   state.CandidateList
+		bCtx             = protocol.MustGetBlockCtx(ctx)
+		fCtx             = protocol.MustGetFeatureCtx(ctx)
+		hCtx             = protocol.MustGetFeatureWithHeightCtx(ctx)
+		numCandDelegates = hCtx.CandDelegateSize(bCtx.BlockHeight)
+	)
 	for i, candidate := range candidates {
-		if uint64(i) >= sh.numCandidateDelegates {
-			break
+		if fCtx.ExtendCandidateDelegatesTo48 {
+			if len(blockProducers) >= int(numCandDelegates) {
+				break
+			}
+		} else {
+			if uint64(i) >= sh.numCandidateDelegates {
+				break
+			}
```

Comment:
simplified to
```
			if len(blockProducers) >= int(numCandDelegates) {
				break
			}
```

Reply:
for the `else` part, if it hits L567 once, the size of  `blockProducers` will be 35, that's why I change to `if` part, so the size will always be 36 (or 48 after extension)

we can try to remove the `else` part after the hard-fork